### PR TITLE
feat: setup ingress for Kind clusters

### LINKED ISSUE #1675: kind: setup ingress or load balancer out of the box
### Is your enhancement related to a problem? Please describe

For now, kind is setup a basic cluster.

Most of time, when deploying application on a cluster, we need to reach the service externally.
But by default, without port-forwarding or ingress controller, etc. it is not possible

### Describe the solution you'd like

We could have ways to have kind being able to setup automatically some optional features

Like: if I check : I want ingress option it'll do https://kind.sigs.k8s.io/docs/user/ingress/
If I check 'LoadBalancer' it could create a metallb configuration https://kind.sigs.k8s.io/docs/user/loadbalancer/

I could also be asked to setup some port forwarding, etc.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jeffmaury**
Fixes #1675

### What does this PR do?

Allow to create a Kind cluster with an ingress controller (Contour)

### Screenshot/screencast of this PR

kind-ingress

### What issues does this PR fix or reference?

Fixes #1675

### How to test this PR?

1. Create a Kind cluster with the Ingress controller option
2. Run `kubectl apply -f https://projectcontour.io/examples/httpbin.yaml`
3. Open http://localhost:9090

### CODE REVIEW

File: extensions/kind/scripts/download.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+if (process.env.GITHUB_TOKEN) {
+  octokitOptions.auth = process.env.GITHUB_TOKEN;
+}
+const octokit = new Octokit(octokitOptions);
+
+// to make this file a module
+export {};
+
+async function download(tagVersion: string, repoPath: string, fileName: string): Promise<void> {
+  const destDir = path.resolve(__dirname, '..', 'src');
```
Comment (benoitf) [inline extensions/kind/scripts/download.ts]: I was wondering if we should download it directly in src folder
I think usually this kind of download is in a specific folder with a gitignore + folder that is removed with `clean` tasks
Might be interesting to use `generated` or `src/generated` (or another convention) just not consider it's already part of the main `src` folder

File: extensions/kind/package.json [hunk 2]
```diff
@@ -41,6 +41,12 @@
           "default": 9443,
           "scope": "KubernetesProviderConnectionFactory",
           "description": "HTTPS Port"
+        },
+        "kind.cluster.creation.ingress": {
+          "type": "boolean",
+          "default": true,
+          "scope": "KubernetesProviderConnectionFactory",
+          "description": "Setup an ingress controller"
```
Comment (benoitf) [inline extensions/kind/package.json]: maybe say that it'll be contour ? (like within parenthesis) 

Also is that mid term we will offer a drop-down with more than Contour ?

**Review by benoitf — COMMENTED** [review]
we could already merge in a separate PR
extension-api: new CreateResource method
implementation in extensionLoader/Kubernetes-client

File: extensions/kind/scripts/download.ts (hunk 1 above)
Comment (jeffmaury) [reply extensions/kind/scripts/download.ts]: Added a new src-generated folder

File: extensions/kind/package.json (hunk 2 above)
Comment (jeffmaury) [reply extensions/kind/package.json]: Updated property description

File: extensions/kind/vitest.config.js [hunk 3]
```diff
@@ -30,6 +31,7 @@ const config = {
   resolve: {
     alias: {
       '@podman-desktop/api': path.resolve('../../', '__mocks__/@podman-desktop/api.js'),
+      '/@gen/': join(PACKAGE_ROOT, 'src-generated') + '/',
```
Comment (benoitf) [inline extensions/kind/vitest.config.js:34]: I'm curious why we need it ? (the one in vite.config.js I understand but not this one)

Comment (jeffmaury) [reply extensions/kind/vitest.config.js:34]: This is to use import of the YAML files in the code without assumption about the location of the generated folder

**Review by lstocchi — APPROVED** [review]
Tested. Works fine 🚀

File: extensions/kind/vitest.config.js (hunk 3 above)
Comment (benoitf) [reply extensions/kind/vitest.config.js:34]: I think there is a code smell (not related to this PR) as vitest config should inherit from the vite.config.js resolve
(it's not blocking this PR)

**Comment by benoitf** [comment]
I've tested, it says successful but at the end I have no contour ingress

```
Creating cluster "kind-cluster" ...
 • Ensuring node image (kindest/node:v1.26.3) 🖼  ...
 ✓ Ensuring node image (kindest/node:v1.26.3) 🖼
 • Preparing nodes 📦   ...
 ✓ Preparing nodes 📦 
 • Writing configuration 📜  ...
 ✓ Writing configuration 📜
 • Starting control-plane 🕹️  ...
 ✓ Starting control-plane 🕹️
 • Installing CNI 🔌  ...
 ✓ Installing CNI 🔌
 • Installing StorageClass 💾  ...
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind-cluster"
You can now use your cluster with:

kubectl cluster-info --context kind-kind-cluster

Thanks for using kind! 😊
```
https://user-images.githubusercontent.com/436777/231242663-023e51d4-e69a-4896-89a0-440b7def7a54.mp4



```shell
kubectl get pods --all-namespaces                                                                                                                                                            NAMESPACE            NAME                                                 READY   STATUS    RESTARTS   AGE
kube-system          coredns-787d4945fb-b7tjz                             1/1     Running   0          99s
kube-system          coredns-787d4945fb-vgfr2                             1/1     Running   0          99s
kube-system          etcd-kind-cluster-control-plane                      1/1     Running   0          111s
kube-system          kindnet-nn9sv                                        1/1     Running   0          26s
kube-system          kube-apiserver-kind-cluster-control-plane            1/1     Running   0          111s
kube-system          kube-controller-manager-kind-cluster-control-plane   1/1     Running   0          111s
kube-system          kube-proxy-f5hmf                                     1/1     Running   0          99s
kube-system          kube-scheduler-kind-cluster-control-plane            1/1     Running   0          111s
local-path-storage   local-path-provisioner-75f5b54ffd-pfj8h              1/1     Running   0          25s
```

**Comment by benoitf** [comment]
It looks like the issue is that it tries to do something on the kind cluster while it's not yet ready (containers are starting) then it gets a connection refused but as there is another issue that does not display the error we see the 'successful message'

so probably we need to wait that the 'kind pod' is ready before applying the contour deployments else we'll probably have some race conditions

**Comment by jeffmaury** [comment]
> It looks like the issue is that it tries to do something on the kind cluster while it's not yet ready (containers are starting) then it gets a connection refused but as there is another issue that does not display the error we see the 'successful message'
> 
> so probably we need to wait that the 'kind pod' is ready before applying the contour deployments else we'll probably have some race conditions

The ingress controller is deployed after the cluster has been created by kind so this should be ok. Do you see the projectcontour namespace ?

**Comment by benoitf** [comment]
no, there is no contour stuff (namespace, pod, crd, etc)

**Comment by jeffmaury** [comment]
> no, there is no contour stuff (namespace, pod, crd, etc)

Looks like the option has not been selected, checking again

**Comment by benoitf** [comment]
@jeffmaury if you could rebase the PR to have all fixes included and merge conflicts solved

**Comment by benoitf** [comment]
The error I got (that is still silently handled) is

```
[kind] Error: No active cluster!
    at KubeConfig.makeApiClient (/Users/benoitf/git/containers/podman-desktop-2/node_modules/@kubernetes/client-node/dist/config.js:298:19)
    at KubernetesClient.createResources (/Users/benoitf/git/containers/podman-desktop-2/packages/main/dist/index.cjs:58194:30)
    at Object.createResources (/Users/benoitf/git/containers/podman-desktop-2/packages/main/dist/index.cjs:37962:33)
    at setupIngressController (/Users/benoitf/git/containers/podman-desktop-2/extensions/kind/dist/extension.js:98064:46)
    at createCluster (/Users/benoitf/git/containers/podman-desktop-2/extensions/kind/dist/extension.js:98106:13)
```

**Comment by benoitf** [comment]
Fixing the error reporting, it's now correctly displaying the error

https://user-images.githubusercontent.com/436777/231697731-9f959ef0-aa3c-41ff-ae0b-dedaf0d735b2.mp4

**Comment by benoitf** [comment]
with the latest commit with kubeConfig path, now I'm able to create a cluster and contour resources have been applied 

```
projectcontour       contour-74866bdd99-kdz76                             0/1     ContainerCreating   0          26s
projectcontour       contour-74866bdd99-x5npc                             0/1     ContainerCreating   0          25s
projectcontour       contour-certgen-v1.24.2-cw7d9                        0/1     Completed           0          26s
projectcontour       envoy-rv8t7                                          0/2     PodInitializing     0          18s
```

not for this PR but probably we should wait if contour is successfully available/running before saying creation is completed/finished

**Review by benoitf — APPROVED** [review]
works now

File: extensions/kind/src/create-cluster.ts [hunk 4]
```diff
@@ -85,6 +115,10 @@ export async function createCluster(
   // now execute the command to create the cluster
   try {
     await runCliCommand(kindCli, ['create', 'cluster', '--config', tmpFilePath], { env, logger }, token);
+    if (ingressController) {
+      logger.log('Creating ingress controller resources');
+      await setupIngressController(clusterName);
+    }
```
Comment (benoitf) [inline extensions/kind/src/create-cluster.ts:121]: probably log into the telemetry logger (few lines after) the value of (ingressController)

**Comment by cdrage** [comment]
After extensive testing, this works, but there are a few things that are not setup 100% afterwards.

The standard Counter example will work, but hostPort forwarding as well as the standard Ingress example on the kind site do not work.

This all LGTM but I've opened up a few issues so we don't forget:

1. Why port 9090? https://github.com/containers/podman-desktop/issues/2073 
2. Code is missing an Ingress patch to setup hostPorts correctly https://kind.sigs.k8s.io/docs/user/ingress/#contour  https://github.com/containers/podman-desktop/issues/2074
3. Default example does not work: https://kind.sigs.k8s.io/docs/user/ingress/#using-ingress (unable to access localhost:9090/foo/hostname https://github.com/containers/podman-desktop/issues/2075

We could fix these issues after as they do not break basic functionality, but other use cases may not work.

**Review by cdrage — APPROVED** [review]

**Pull request merged**